plugins/nf-prov/src/main/nextflow/prov/BcoRenderer.groovy

```diff
@@ -0,0 +1,238 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package nextflow.prov
+
+import java.nio.file.Path
+import java.time.format.DateTimeFormatter
+
+import groovy.json.JsonOutput
+import groovy.transform.CompileStatic
+import nextflow.Session
+import nextflow.exception.AbortOperationException
+import nextflow.processor.TaskRun
+import nextflow.script.WorkflowMetadata
+import nextflow.script.params.FileOutParam
+import nextflow.util.CacheHelper
+
+/**
+ * Renderer for the BioCompute Object (BCO) format.
+ *
+ * @author Ben Sherman <bentshermann@gmail.com>
+ */
+@CompileStatic
+class BcoRenderer implements Renderer {
+
+    private WorkflowMetadata metadata
+
+    /**
+     * Get the list of output files for a task.
+     *
+     * @param task
+     */
+    private static List<Path> getTaskOutputs(TaskRun task) {
+        return task
+            .getOutputsByType(FileOutParam)
+            .values()
+            .flatten() as List<Path>
+    }
+
+    /**
+     * Get a mapping of output file to the task that produced it.
+     *
+     * @param tasks
+     */
+    private static Map<Path,TaskRun> getTaskLookup(Set<TaskRun> tasks) {
+        final taskLookup = [:] as Map<Path,TaskRun>
+
+        for( def task : tasks )
+            for( def output : getTaskOutputs(task) )
+                taskLookup[output] = task
+
+        return taskLookup
+    }
+
+    /**
+     * Get the list of workflow inputs. A workflow input is an input file
+     * to a task that was not produced by another task.
+     *
+     * @param tasks
+     * @param taskLookup
+     */
+    private static Set<Path> getWorkflowInputs(Set<TaskRun> tasks, Map<Path,TaskRun> taskLookup) {
+        final workflowInputs = [] as Set<Path>
+
+        tasks.each { task ->
+            task.getInputFilesMap().each { name, path ->
+                if( taskLookup[path] )
+                    return
+
+                workflowInputs << path
+            }
+        }
+
+        return workflowInputs
+    }
+
+    /**
+     * Normalize local paths to remove environment-specific directories.
+     *
+     * @param path
+     */
+    private String normalizePath(Path path) {
+        // TODO: append raw github URL to git assets
+        path.toUriString()
+            .replace(metadata.workDir.toUriString(), '/work')
+            .replace(metadata.projectDir.toUriString(), '')
+            .replace(metadata.launchDir.toUriString(), '')
+    }
+
+    @Override
+    void render(Session session, Set<TaskRun> tasks, Map<Path,Path> workflowOutputs, Path path) {
+        if( !path.mkdirs() )
```

Comment:
Rename any existing folder to `<name>.bak` or similar instead of failing on an existing folder?

Reply:
This line will fail only if the folder couldn't be created, not if it already exists